File: Models/SVM.py
import numpy as np
import matplotlib.pyplot as plt
from tensorflow.keras.datasets import mnist # type: ignore
from sklearn.svm import SVC
from sklearn.decomposition import PCA
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
import seaborn as sns
import time

# Load MNIST dataset
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

# Normalize pixel values to [0, 1]
train_images = train_images / 255.0
test_images = test_images / 255.0

# Flatten images
train_images_flat = train_images.reshape(train_images.shape[0], -1)
test_images_flat = test_images.reshape(test_images.shape[0], -1)

# Apply PCA to reduce dimensionality (Ensure PCA is fitted ONLY on training data)
pca = PCA(n_components=50, svd_solver='randomized', whiten=True)  # Retain 95%+ variance
pca.fit(train_images_flat)  # Fit only on train data
train_images_pca = pca.transform(train_images_flat)
test_images_pca = pca.transform(test_images_flat)  # Apply same transformation to test data

# Compute the average number of activated pixels per digit (nonzero pixels)
average_active_pixels = {
    digit: np.mean([np.count_nonzero(img) for img in train_images[train_labels == digit]])
    for digit in range(10)
}

# Initialize SVM with RBF kernel and prevent overfitting
model_svm = SVC(kernel='rbf', gamma=0.005, C=0.5, verbose=True)  # Reduced gamma and C

# Measure training time
print("Training SVM...")
start_time = time.time()
model_svm.fit(train_images_pca, train_labels)
training_time = time.time() - start_time
print(f"Training completed in {training_time:.2f} seconds.")

# Check number of support vectors to diagnose overfitting
print(f"Number of Support Vectors per class: {model_svm.n_support_}")

# Measure inference time
print("Running predictions...")
start_time = time.time()
svm_predictions = model_svm.predict(test_images_pca)
inference_time = time.time() - start_time
print(f"Inference completed in {inference_time:.2f} seconds.")

# Evaluate accuracy
svm_accuracy = accuracy_score(test_labels, svm_predictions)
print(f"SVM Accuracy: {svm_accuracy:.4f}")

# Debugging: Ensure the predictions are reasonable
print("First 10 Test Labels: ", test_labels[:10])
print("First 10 Predictions: ", svm_predictions[:10])

# Print the average number of activated pixels for each digit
print("\nAverage Number of Activated Pixels for Each Digit:")
for digit, avg_pixels in average_active_pixels.items():
    print(f"Digit {digit}: {avg_pixels:.2f} pixels")

# Confusion Matrix (to identify frequently misclassified digits)
cm = confusion_matrix(test_labels, svm_predictions)
plt.figure(figsize=(10, 8))
sns.heatmap(cm, annot=True, fmt='d', cmap='Blues', cbar=False)
plt.xlabel('Predicted')
plt.ylabel('True')
plt.title('Confusion Matrix for SVM')
plt.show()

# Classification Report (precision, recall, F1-score)
print("Classification Report:")
print(classification_report(test_labels, svm_predictions))

# Visualize misclassified digits for each class
misclassified = np.where(test_labels != svm_predictions)[0]
misclassified_by_class = {i: [] for i in range(10)}

# Group misclassified samples by true label
for idx in misclassified:
    true_label = test_labels[idx]
    misclassified_by_class[true_label].append(idx)

# Plot one misclassified example per class
plt.figure(figsize=(15, 8))
for i in range(10):
    if misclassified_by_class[i]:  # Check if there are misclassified samples for this class
        idx = misclassified_by_class[i][0]  # Take the first misclassified sample for this class
        plt.subplot(2, 5, i+1)
        plt.imshow(test_images[idx], cmap='gray')
        plt.title(f"True: {test_labels[idx]}\nPred: {svm_predictions[idx]}")
    else:
        plt.subplot(2, 5, i+1)
        plt.text(0.5, 0.5, f"No misclassified\nfor {i}", ha='center', va='center')
        plt.axis('off')
plt.suptitle('Misclassified Digits by SVM (One Example per Class)')
plt.show()

# Calculate misclassified counts by digit
misclassified_counts = {i: 0 for i in range(10)}
for idx in misclassified:
    true_label = test_labels[idx]
    misclassified_counts[true_label] += 1

# Calculate successful classification percentage by digit
success_percentage = {i: 0 for i in range(10)}
for i in range(10):
    total_samples = np.sum(test_labels == i)
    correct_samples = total_samples - misclassified_counts[i]
    success_percentage[i] = (correct_samples / total_samples) * 100

# Print results
print(f"SVM Model Training Time: {training_time:.2f} seconds")
print(f"SVM Model Testing Time: {inference_time:.2f} seconds")
print(f"SVM Model Test Accuracy: {svm_accuracy * 100:.2f}%")
print("\nMisclassified counts by digit for SVM:")
for i in range(10):
    print(f"Digit {i}: {misclassified_counts[i]}")
print("\nSuccessful classification percentage for SVM by digit:")
for i in range(10):
    print(f"Digit {i}: {success_percentage[i]:.2f}%")

File: Outputs/SVM/SVM.rs
Training SVM...
*
optimization finished, #iter = 423
obj = -47.429628, rho = -0.333603
nSV = 190, nBSV = 34
*.*
...
... /SVM/Optamizations.rs
...
[LibSVM]Training completed in 33.14 seconds.
Running predictions...
Inference completed in 18.20 seconds.
SVM Accuracy: 0.9833

Average Number of Activated Pixels for Each Digit:
Digit 0: 191.97 pixels
Digit 1: 85.85 pixels
Digit 2: 168.81 pixels
Digit 3: 163.34 pixels
Digit 4: 141.81 pixels
Digit 5: 152.33 pixels
Digit 6: 156.93 pixels
Digit 7: 131.40 pixels
Digit 8: 173.32 pixels
Digit 9: 143.03 pixels

2025-02-03 15:27:19.866 Python[74853:9506377] +[IMKClient subclass]: chose IMKClient_Modern
2025-02-03 15:27:19.866 Python[74853:9506377] +[IMKInputSession subclass]: chose IMKInputSession_Modern
Classification Report:
              precision    recall  f1-score   support

           0       0.98      0.99      0.99       980
           1       0.99      0.99      0.99      1135
           2       0.98      0.98      0.98      1032
           3       0.98      0.99      0.98      1010
           4       0.98      0.98      0.98       982
           5       0.99      0.98      0.98       892
           6       0.99      0.98      0.99       958
           7       0.98      0.98      0.98      1028
           8       0.98      0.99      0.98       974
           9       0.98      0.97      0.97      1009

    accuracy                           0.98     10000
   macro avg       0.98      0.98      0.98     10000
weighted avg       0.98      0.98      0.98     10000

SVM Model Training Time: 33.14 seconds
SVM Model Testing Time: 18.20 seconds
SVM Model Test Accuracy: 98.33%

Misclassified counts by digit for SVM:
Digit 0: 6
Digit 1: 6
Digit 2: 21
Digit 3: 15
Digit 4: 17
Digit 5: 16
Digit 6: 15
Digit 7: 25
Digit 8: 14
Digit 9: 32

Successful classification percentage for SVM by digit:
Digit 0: 99.39%
Digit 1: 99.47%
Digit 2: 97.97%
Digit 3: 98.51%
Digit 4: 98.27%
Digit 5: 98.21%
Digit 6: 98.43%
Digit 7: 97.57%
Digit 8: 98.56%
Digit 9: 96.83%